internal/rest/resources/control.go

```diff
@@ -75,12 +75,26 @@ func controlPost(state state.State, r *http.Request) response.Response {
 		return joinWithToken(state, r, req)
 	}
 
+	reverter := revert.New()
```

Comment:
Do you set the reverter before calling `StartAPI()` so that if the call fails the revert gets executed?

To keep the common pattern of first performing an action and then registering the reverter afterwards, I am wondering if the reverter would be better placed inside `StartAPI()` like it's done in `joinWithToken()`?

Reply:
`StartAPI` is also called when the daemon restarts, where a failure doesn't require us to reset the whole daemon. It's only in the bootstrap and join cases where we have to reset the daemon back to its initial state.

Reply:
In this case would it make sense to return a `cleanup` hook from `StartAPI()`? So the caller can decide whether or not to run the revert.

Reply:
I'm not sure if that would make things unclear/dangerous. The re-exec logic does not "belong" to `StartAPI`, which may eventually be used in more  situations and require situational cleanup. We don't necessarily want to elicit the question of "do you want to erase all data in the state directory and forcibly kill the daemon" for each call to `StartAPI`.

I think this revert is in line with the pattern you mentioned, it is tied to the control API endpoint used for bootstrap/join. The "action" in this case is handling that API endpoint.

Reply:
Thinking a bit more about it now I agree that we shouldn't make it more complex. 

It's just that you register the revert (`resetClusterMember()`) before the actual actions have been taken as part of `StartAPI()`.
Is `resetClusterMember()` idempotent in a way that it doesn't matter how far into `StartAPI()` an error is returned?

Reply:
> Is `resetClusterMember()` idempotent in a way that it doesn't matter how far into `StartAPI()` an error is returned?

I'm not fully following you, can you give an example of what you mean here?

`resetClusterMember` attempts to stop the database and delete some files in case they were created. We run it with `force=true` so it will always try to fully clean up and ignore errors. As such I don't think the timing with `StartAPI` is of much concern.

Reply:
Yeah that's what I meant. I missed the fact that we have a `force` flag for this func which is set to `true` in this case.
Fine for me then.